cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/BatchSparkPipelineDriver.java

```diff
@@ -118,8 +119,8 @@ protected BatchSparkPipelineDriver(BatchSQLEngineAdapter sqlEngineAdapter) {
 
   @Override
   protected SparkCollection<RecordInfo<Object>> getSource(StageSpec stageSpec,
-                                                          FunctionCache.Factory functionCacheFactory,
-                                                          StageStatisticsCollector collector) {
+      FunctionCache.Factory functionCacheFactory, StageStatisticsCollector collector,
+      Callable<Void> batchRetryFunction) {
```

Comment:
is there any way to avoid passing this Callable around everywhere? It's a bit odd here since it's not ever used for batch pipelines, but it's still passed around for them.

Reply:
Extracted the sink function methods to an interface so that the retry callable need not be passed to all the methods.